junit-jupiter-api/src/main/java/org/junit/jupiter/api/MethodOrderer.java

```diff
@@ -172,6 +175,76 @@ private static int getOrder(MethodDescriptor descriptor) {
 		}
 	}
 
+	/**
+	 * {@code MethodOrderer} that sorts methods based on the {@link Depend @Depend}
+	 * annotation.
+	 */
+	class DependAnnotation implements MethodOrderer {
+		private static final Logger logger = LoggerFactory.getLogger(Random.class);
+
+		private static Map<String, Integer> dependencySize;
+		private static Map<String, String[]> digraph;
+
+		@Override
+		public void orderMethods(MethodOrdererContext context) {
+			digraph = context.getMethodDescriptors().stream().filter(
+				descriptor -> descriptor.findAnnotation(Depend.class).isPresent()).collect(
+					Collectors.toMap(descriptor -> descriptor.getMethod().getName(),
+						descriptor -> descriptor.findAnnotation(Depend.class).map(Depend::methods).get()));
+
+			dependencySize = new HashMap<>();
+
+			try {
+				for (String name : digraph.keySet()) {
+					if (!dependencySize.containsKey(name)) {
+						int sz = DFS(name);
+						if (sz == -1) {
+							throw new IllegalArgumentException("Cycle Detected!!!");
+						}
+					}
+				}
+			}
+			catch (IllegalArgumentException exception) {
+				logger.error(exception,
+					() -> "ERROR - Some arguments from @Depend annotations form cyclic dependencies, which would cause undefined behavior!");
+			}
+
+			context.getMethodDescriptors().sort(Comparator.comparing(DependAnnotation::getDependencySize));
+		}
+
+		private static int getDependencySize(MethodDescriptor descriptor) {
+			return dependencySize.getOrDefault(descriptor.getMethod().getName(), 0);
+		}
+
+		private static int DFS(String name) {
+			if (dependencySize.containsKey(name)) {
+				return dependencySize.get(name);
+			}
+			// mark entering this node
+			dependencySize.put(name, -1);
+
+			String[] ancestors = digraph.get(name);
+			int total = 1;
+
+			if (ancestors != null) {
+				for (String ancestor : ancestors) {
+					int sz = DFS(ancestor);
+
+					// cycle detected
+					if (sz == -1) {
+						return -1;
+					}
```

Comment:
Now that I've looked more closely at the code, it appears to me that it would be better to simply throw an exception with an informative message at the first detection of a cycle, like this:

```java
if (sz == -1) {
	throw new IllegalArgumentException(
		String.format("Cycle detected between %s and %s", name, ancestor));
}
```

Reply:
I have edited as suggested.